File: skills/agent-patterns/templates/python/parallelization.py
"""
Parallelization Pattern Implementation
Execute independent tasks concurrently (Sectioning) or same task multiple times (Voting)
"""

from typing import Any, Callable, Optional
from dataclasses import dataclass
import asyncio
import anthropic
from collections import Counter


@dataclass
class Section:
    """Represents a parallel subtask"""
    name: str
    prompt: str
    data: Optional[Any] = None


@dataclass
class VoteResult:
    """Result of voting aggregation"""
    consensus: str
    votes: list[str]
    confidence: float
    vote_counts: dict[str, int]


class SectioningParallelizer:
    """
    Break task into independent subtasks and execute concurrently.
    
    Use when:
    - Subtasks are truly independent (no dependencies)
    - Speed/throughput is important
    - Results can be meaningfully combined
    
    Example:
        parallelizer = SectioningParallelizer(client)
        sections = [
            Section("security", "Analyze code for security vulnerabilities", code),
            Section("performance", "Analyze code for performance issues", code),
            Section("style", "Check code style and best practices", code),
        ]
        results = await parallelizer.execute(sections)
    """
    
    def __init__(
        self,
        client: anthropic.Anthropic,
        model: str = "claude-sonnet-4-20250514"
    ):
        self.client = client
        self.model = model
        self.results: dict[str, str] = {}
    
    async def _execute_section(self, section: Section) -> tuple[str, str]:
        """Execute a single section"""
        prompt = section.prompt
        if section.data:
            prompt = f"{section.prompt}\n\nContent:\n{section.data}"
        
        message = await self.client.messages.create(
            model=self.model,
            max_tokens=4096,
            messages=[{"role": "user", "content": prompt}]
        )
        
        return section.name, message.content[0].text
    
    async def execute(self, sections: list[Section]) -> dict[str, str]:
        """Execute all sections in parallel"""
        tasks = [self._execute_section(section) for section in sections]
        results = await asyncio.gather(*tasks)
        
        self.results = dict(results)
        return self.results
    
    async def execute_and_combine(
        self,
        sections: list[Section],
        combine_prompt: str
    ) -> str:
        """Execute sections in parallel and combine results"""
        results = await self.execute(sections)
        
        # Format results for combination
        results_text = "\n\n".join([
            f"## {name}\n{result}"
            for name, result in results.items()
        ])
        
        combine_message = await self.client.messages.create(
            model=self.model,
            max_tokens=4096,
            messages=[{
                "role": "user",
                "content": f"{combine_prompt}\n\n{results_text}"
            }]
        )
        
        return combine_message.content[0].text


class VotingParallelizer:
    """
    Run the same task multiple times and aggregate for robustness.
    
    Use when:
    - Critical accuracy is needed
    - Consensus improves quality
    - Different prompts/approaches provide diverse perspectives
    - Cost of errors significantly exceeds compute cost
    
    Example:
        voter = VotingParallelizer(client)
        result = await voter.vote(
            "Is this code safe to deploy?",
            code_content,
            num_votes=3
        )
        print(f"Consensus: {result.consensus} ({result.confidence:.0%} confidence)")
    """
    
    def __init__(
        self,
        client: anthropic.Anthropic,
        model: str = "claude-sonnet-4-20250514"
    ):
        self.client = client
        self.model = model
    
    async def _get_vote(
        self,
        prompt: str,
        content: str,
        vote_id: int
    ) -> str:
        """Get a single vote"""
        # Slightly vary temperature or add random seed for diversity
        message = await self.client.messages.create(
            model=self.model,
            max_tokens=1024,
            messages=[{
                "role": "user",
                "content": f"{prompt}\n\nContent:\n{content}\n\nProvide your answer as a single word or short phrase."
            }]
        )
        
        return message.content[0].text.strip()
    
    async def vote(
        self,
        prompt: str,
        content: str,
        num_votes: int = 3
    ) -> VoteResult:
        """Run voting with multiple LLM calls"""
        tasks = [
            self._get_vote(prompt, content, i)
            for i in range(num_votes)
        ]
        votes = await asyncio.gather(*tasks)
        
        # Count votes (normalize to lowercase for comparison)
        normalized = [v.lower() for v in votes]
        vote_counts = Counter(normalized)
        
        # Get consensus (most common vote)
        consensus, count = vote_counts.most_common(1)[0]
        confidence = count / num_votes
        
        return VoteResult(
            consensus=consensus,
            votes=list(votes),
            confidence=confidence,
            vote_counts=dict(vote_counts)
        )
    
    async def vote_with_perspectives(
        self,
        prompts: list[str],
        content: str
    ) -> VoteResult:
        """Vote using different prompts/perspectives"""
        tasks = [
            self._get_vote(prompt, content, i)
            for i, prompt in enumerate(prompts)
        ]
        votes = await asyncio.gather(*tasks)
        
        normalized = [v.lower() for v in votes]
        vote_counts = Counter(normalized)
        consensus, count = vote_counts.most_common(1)[0]
        
        return VoteResult(
            consensus=consensus,
            votes=list(votes),
            confidence=count / len(prompts),
            vote_counts=dict(vote_counts)
        )


class GuardrailsParallelizer:
    """
    Run guardrails in parallel with main task.
    
    Example from Anthropic blog: One model processes user queries while
    another screens them for inappropriate content. This performs better
    than having the same LLM handle both.
    """
    
    def __init__(
        self,
        client: anthropic.Anthropic,
        model: str = "claude-sonnet-4-20250514"
    ):
        self.client = client
        self.model = model
    
    async def _check_safety(self, content: str) -> dict[str, Any]:
        """Check content for safety issues"""
        message = await self.client.messages.create(
            model="claude-3-5-haiku-20241022",  # Fast model for safety check
            max_tokens=256,
            system="""You are a content safety classifier. Analyze the content for:
- Harmful content
- Inappropriate requests  
- Policy violations

Respond with JSON:
{
    "safe": true/false,
    "reason": "explanation if unsafe"
}""",
            messages=[{"role": "user", "content": content}]
        )
        
        import json
        response_text = message.content[0].text
        if "```json" in response_text:
            response_text = response_text.split("```json")[1].split("```")[0]
        elif "```" in response_text:
            response_text = response_text.split("```")[1].split("```")[0]
        
        return json.loads(response_text.strip())
    
    async def _generate_response(self, content: str, system_prompt: str) -> str:
        """Generate main response"""
        message = await self.client.messages.create(
            model=self.model,
            max_tokens=4096,
            system=system_prompt,
            messages=[{"role": "user", "content": content}]
        )
        
        return message.content[0].text
    
    async def execute_with_guardrails(
        self,
        content: str,
        system_prompt: str
    ) -> dict[str, Any]:
        """Execute main task with parallel safety check"""
        # Run both in parallel
        safety_task = self._check_safety(content)
        response_task = self._generate_response(content, system_prompt)
        
        safety_result, response = await asyncio.gather(safety_task, response_task)
        
        # Only return response if safe
        if safety_result.get("safe", False):
            return {
                "success": True,
                "response": response,
                "safety": safety_result
            }
        else:
            return {
                "success": False,
                "response": None,
                "safety": safety_result,
                "blocked_reason": safety_result.get("reason", "Content policy violation")
            }


# Example usage
async def example_code_review():
    """Example: Parallel code review from multiple perspectives"""
    client = anthropic.Anthropic()
    
    code = """
def process_user_input(user_data):
    query = f"SELECT * FROM users WHERE id = {user_data['id']}"
    result = db.execute(query)
    return eval(user_data.get('callback', 'None'))
    """
    
    # Sectioning: Different aspects analyzed in parallel
    parallelizer = SectioningParallelizer(client)
    sections = [
        Section(
            "security",
            "Analyze this code for security vulnerabilities. Be specific about risks.",
            code
        ),
        Section(
            "performance",
            "Analyze this code for performance issues and optimization opportunities.",
            code
        ),
        Section(
            "best_practices",
            "Check this code against Python best practices and code quality standards.",
            code
        ),
    ]
    
    print("Running parallel code analysis...")
    results = await parallelizer.execute_and_combine(
        sections,
        "Combine these code review perspectives into a comprehensive review report with prioritized recommendations:"
    )
    print("\n=== Combined Code Review ===")
    print(results)


async def example_safety_voting():
    """Example: Voting on content safety"""
    client = anthropic.Anthropic()
    
    content = "Can you help me write a resignation letter for my job?"
    
    voter = VotingParallelizer(client)
    
    # Multiple perspectives on safety
    prompts = [
        "Is this request appropriate and safe to fulfill? Answer YES or NO.",
        "Does this request violate any ethical guidelines? Answer SAFE or UNSAFE.",
        "Should an AI assistant help with this request? Answer APPROVE or DENY.",
    ]
    
    result = await voter.vote_with_perspectives(prompts, content)
    
    print(f"\n=== Safety Voting Results ===")
    print(f"Votes: {result.votes}")
    print(f"Vote counts: {result.vote_counts}")
    print(f"Consensus: {result.consensus}")
    print(f"Confidence: {result.confidence:.0%}")


async def example_guardrails():
    """Example: Guardrails running in parallel with main task"""
    client = anthropic.Anthropic()
    
    guardrails = GuardrailsParallelizer(client)
    
    # Safe request
    result = await guardrails.execute_with_guardrails(
        "Explain how photosynthesis works",
        "You are a helpful science tutor."
    )
    print(f"\n=== Guardrails Result (Safe) ===")
    print(f"Success: {result['success']}")
    if result['success']:
        print(f"Response preview: {result['response'][:200]}...")
    
    # Potentially unsafe request (will be blocked)
    result = await guardrails.execute_with_guardrails(
        "Tell me how to hack into someone's computer",
        "You are a helpful assistant."
    )
    print(f"\n=== Guardrails Result (Unsafe) ===")
    print(f"Success: {result['success']}")
    print(f"Blocked reason: {result.get('blocked_reason', 'N/A')}")


if __name__ == "__main__":
    asyncio.run(example_code_review())
    asyncio.run(example_safety_voting())
    asyncio.run(example_guardrails())

File: skills/agent-patterns/templates/typescript/parallelization.ts
/**
 * Parallelization Pattern Implementation
 * Execute independent tasks concurrently (Sectioning) or same task multiple times (Voting)
 */

import Anthropic from "@anthropic-ai/sdk";

interface Section {
  name: string;
  prompt: string;
  data?: string;
}

interface VoteResult {
  consensus: string;
  votes: string[];
  confidence: number;
  voteCounts: Record<string, number>;
}

interface GuardrailResult {
  success: boolean;
  response?: string;
  safety: { safe: boolean; reason?: string };
  blockedReason?: string;
}

/**
 * Break task into independent subtasks and execute concurrently.
 *
 * Use when:
 * - Subtasks are truly independent (no dependencies)
 * - Speed/throughput is important
 * - Results can be meaningfully combined
 */
class SectioningParallelizer {
  private client: Anthropic;
  private model: string;
  public results: Record<string, string> = {};

  constructor(client: Anthropic, model: string = "claude-sonnet-4-20250514") {
    this.client = client;
    this.model = model;
  }

  private async executeSection(section: Section): Promise<[string, string]> {
    let prompt = section.prompt;
    if (section.data) {
      prompt = `${section.prompt}\n\nContent:\n${section.data}`;
    }

    const message = await this.client.messages.create({
      model: this.model,
      max_tokens: 4096,
      messages: [{ role: "user", content: prompt }],
    });

    const text =
      message.content[0].type === "text" ? message.content[0].text : "";
    return [section.name, text];
  }

  async execute(sections: Section[]): Promise<Record<string, string>> {
    const promises = sections.map((section) => this.executeSection(section));
    const results = await Promise.all(promises);

    this.results = Object.fromEntries(results);
    return this.results;
  }

  async executeAndCombine(
    sections: Section[],
    combinePrompt: string
  ): Promise<string> {
    const results = await this.execute(sections);

    // Format results for combination
    const resultsText = Object.entries(results)
      .map(([name, result]) => `## ${name}\n${result}`)
      .join("\n\n");

    const combineMessage = await this.client.messages.create({
      model: this.model,
      max_tokens: 4096,
      messages: [
        {
          role: "user",
          content: `${combinePrompt}\n\n${resultsText}`,
        },
      ],
    });

    return combineMessage.content[0].type === "text"
      ? combineMessage.content[0].text
      : "";
  }
}

/**
 * Run the same task multiple times and aggregate for robustness.
 *
 * Use when:
 * - Critical accuracy is needed
 * - Consensus improves quality
 * - Different prompts/approaches provide diverse perspectives
 * - Cost of errors significantly exceeds compute cost
 */
class VotingParallelizer {
  private client: Anthropic;
  private model: string;

  constructor(client: Anthropic, model: string = "claude-sonnet-4-20250514") {
    this.client = client;
    this.model = model;
  }

  private async getVote(
    prompt: string,
    content: string,
    _voteId: number
  ): Promise<string> {
    const message = await this.client.messages.create({
      model: this.model,
      max_tokens: 1024,
      messages: [
        {
          role: "user",
          content: `${prompt}\n\nContent:\n${content}\n\nProvide your answer as a single word or short phrase.`,
        },
      ],
    });

    const text =
      message.content[0].type === "text" ? message.content[0].text : "";
    return text.trim();
  }

  async vote(
    prompt: string,
    content: string,
    numVotes: number = 3
  ): Promise<VoteResult> {
    const promises = Array.from({ length: numVotes }, (_, i) =>
      this.getVote(prompt, content, i)
    );
    const votes = await Promise.all(promises);

    // Count votes (normalize to lowercase)
    const normalized = votes.map((v) => v.toLowerCase());
    const voteCounts: Record<string, number> = {};
    normalized.forEach((v) => {
      voteCounts[v] = (voteCounts[v] || 0) + 1;
    });

    // Get consensus (most common vote)
    const consensus = Object.entries(voteCounts).sort(
      ([, a], [, b]) => b - a
    )[0][0];
    const confidence = voteCounts[consensus] / numVotes;

    return {
      consensus,
      votes,
      confidence,
      voteCounts,
    };
  }

  async voteWithPerspectives(
    prompts: string[],
    content: string
  ): Promise<VoteResult> {
    const promises = prompts.map((prompt, i) =>
      this.getVote(prompt, content, i)
    );
    const votes = await Promise.all(promises);

    const normalized = votes.map((v) => v.toLowerCase());
    const voteCounts: Record<string, number> = {};
    normalized.forEach((v) => {
      voteCounts[v] = (voteCounts[v] || 0) + 1;
    });

    const consensus = Object.entries(voteCounts).sort(
      ([, a], [, b]) => b - a
    )[0][0];

    return {
      consensus,
      votes,
      confidence: voteCounts[consensus] / prompts.length,
      voteCounts,
    };
  }
}

/**
 * Run guardrails in parallel with main task.
 *
 * From Anthropic blog: One model processes user queries while
 * another screens them for inappropriate content. This performs better
 * than having the same LLM handle both.
 */
class GuardrailsParallelizer {
  private client: Anthropic;
  private model: string;

  constructor(client: Anthropic, model: string = "claude-sonnet-4-20250514") {
    this.client = client;
    this.model = model;
  }

  private async checkSafety(
    content: string
  ): Promise<{ safe: boolean; reason?: string }> {
    const message = await this.client.messages.create({
      model: "claude-3-5-haiku-20241022", // Fast model for safety check
      max_tokens: 256,
      system: `You are a content safety classifier. Analyze the content for:
- Harmful content
- Inappropriate requests  
- Policy violations

Respond with JSON:
{
    "safe": true/false,
    "reason": "explanation if unsafe"
}`,
      messages: [{ role: "user", content }],
    });

    let responseText =
      message.content[0].type === "text" ? message.content[0].text : "{}";

    if (responseText.includes("```json")) {
      responseText = responseText.split("```json")[1].split("```")[0];
    } else if (responseText.includes("```")) {
      responseText = responseText.split("```")[1].split("```")[0];
    }

    return JSON.parse(responseText.trim());
  }

  private async generateResponse(
    content: string,
    systemPrompt: string
  ): Promise<string> {
    const message = await this.client.messages.create({
      model: this.model,
      max_tokens: 4096,
      system: systemPrompt,
      messages: [{ role: "user", content }],
    });

    return message.content[0].type === "text" ? message.content[0].text : "";
  }

  async executeWithGuardrails(
    content: string,
    systemPrompt: string
  ): Promise<GuardrailResult> {
    // Run both in parallel
    const [safetyResult, response] = await Promise.all([
      this.checkSafety(content),
      this.generateResponse(content, systemPrompt),
    ]);

    // Only return response if safe
    if (safetyResult.safe) {
      return {
        success: true,
        response,
        safety: safetyResult,
      };
    } else {
      return {
        success: false,
        response: undefined,
        safety: safetyResult,
        blockedReason: safetyResult.reason ?? "Content policy violation",
      };
    }
  }
}

// Example usage
async function exampleCodeReview() {
  const client = new Anthropic();

  const code = `
def process_user_input(user_data):
    query = f"SELECT * FROM users WHERE id = {user_data['id']}"
    result = db.execute(query)
    return eval(user_data.get('callback', 'None'))
  `;

  // Sectioning: Different aspects analyzed in parallel
  const parallelizer = new SectioningParallelizer(client);
  const sections: Section[] = [
    {
      name: "security",
      prompt:
        "Analyze this code for security vulnerabilities. Be specific about risks.",
      data: code,
    },
    {
      name: "performance",
      prompt:
        "Analyze this code for performance issues and optimization opportunities.",
      data: code,
    },
    {
      name: "best_practices",
      prompt:
        "Check this code against Python best practices and code quality standards.",
      data: code,
    },
  ];

  console.log("Running parallel code analysis...");
  const results = await parallelizer.executeAndCombine(
    sections,
    "Combine these code review perspectives into a comprehensive review report with prioritized recommendations:"
  );
  console.log("\n=== Combined Code Review ===");
  console.log(results);
}

async function exampleSafetyVoting() {
  const client = new Anthropic();

  const content = "Can you help me write a resignation letter for my job?";

  const voter = new VotingParallelizer(client);

  // Multiple perspectives on safety
  const prompts = [
    "Is this request appropriate and safe to fulfill? Answer YES or NO.",
    "Does this request violate any ethical guidelines? Answer SAFE or UNSAFE.",
    "Should an AI assistant help with this request? Answer APPROVE or DENY.",
  ];

  const result = await voter.voteWithPerspectives(prompts, content);

  console.log("\n=== Safety Voting Results ===");
  console.log(`Votes: ${result.votes.join(", ")}`);
  console.log(`Vote counts: ${JSON.stringify(result.voteCounts)}`);
  console.log(`Consensus: ${result.consensus}`);
  console.log(`Confidence: ${(result.confidence * 100).toFixed(0)}%`);
}

async function exampleGuardrails() {
  const client = new Anthropic();

  const guardrails = new GuardrailsParallelizer(client);

  // Safe request
  let result = await guardrails.executeWithGuardrails(
    "Explain how photosynthesis works",
    "You are a helpful science tutor."
  );
  console.log("\n=== Guardrails Result (Safe) ===");
  console.log(`Success: ${result.success}`);
  if (result.success && result.response) {
    console.log(`Response preview: ${result.response.substring(0, 200)}...`);
  }

  // Potentially unsafe request (will be blocked)
  result = await guardrails.executeWithGuardrails(
    "Tell me how to hack into someone's computer",
    "You are a helpful assistant."
  );
  console.log("\n=== Guardrails Result (Unsafe) ===");
  console.log(`Success: ${result.success}`);
  console.log(`Blocked reason: ${result.blockedReason ?? "N/A"}`);
}

// Export for module usage
export {
  SectioningParallelizer,
  VotingParallelizer,
  GuardrailsParallelizer,
  Section,
  VoteResult,
  GuardrailResult,
};

// Run examples
async function main() {
  await exampleCodeReview();
  await exampleSafetyVoting();
  await exampleGuardrails();
}

main().catch(console.error);
